### ISSUE TITLE
Build issue: Build picks up wrong git info

### ISSUE THREAD

**Issue by ferdnyc** [doxygen/doxygen#10995]
This is similar to #10643, but coming from a different angle.

**Describe the bug**
I'm building in a Fedora tree where:

1. Our packaging configs are a Git repo, say it's `/src/doxygen/`
2. The package uses `doxygen-1.11.0.src.tar.gz` (the source tarball) as its source
3. Local builds are done by extracting the tarball to `/src/doxygen/doxygen-1.11.0`
4. `git` commands inside the `doxygen-1.11.0` source directory will therefore _work_, and return data from `/src/doxygen/.git`.

**To Reproduce**
```bash
mkdir /tmp/doxygen-test
cd /tmp/doxygen-test
git init
touch dummy-file.txt
git commit -m "Initial commit" dummy-file.txt

# The output here will be something like 'heads/main-0-<shorthash>'
git describe --all --long

wget 'https://www.doxygen.nl/files/doxygen-1.11.0.src.tar.gz'
tar xf doxygen-1.11.0.src.tar.gz
cd doxygen-1.11.0
cmake -B _build -S .
cmake --build _build
```

**Expected behavior**
Doxygen builds with no Git version

**Actual behavior**
`doxygen --version` shows `1.11.0 (<the_hash_from_git_describe_above>*)`

**Version**
Building 1.11.0 from source

**Additional editorializing**
IMHO it would be better to just _look for_ a `.git` directory at the root of the Doxygen source, and disable all Git processing unless one's found. Blindly running `git` commands and making decisions based on whether or not they succeed is far too error-prone.

**Comment by ferdnyc** [comment]
> IMHO it would be better to just _look for_ a `.git` directory at the root of the Doxygen source

Actually, correction — make that: "look for `.git` at the root of the Doxygen source", since in a Git [worktree](https://git-scm.com/docs/git-worktree) there will be a `.git` **file** that points to the parent source tree. Even if `.git` is a file, not a directory, `git` commands from within that tree can be considered valid.

**Comment by albert-github** [comment]
It is a bit a twisted example but it, unfortunately, is possible.

I think in the first block there are a few commands missing (at least as far as I can tell:
- git add dummy-file.txt
- git config --global user.email "you@example.com"
- git config --global user.name "Your Name"

The detection of the git version is based on `cmake/git_watcher.cmake` which is derived from https://github.com/andrew-hardin/cmake-git-version-tracking /  https://raw.githubusercontent.com/andrew-hardin/cmake-git-version-tracking/master/git_watcher.cmake so this should be checked what is possible and do something with `${PROJECT_SOURCE_DIR}/.git`

**Comment by ferdnyc** [comment]
> * git add dummy-file.txt

Not necessary, actually, if you specify the filename(s) in the `git commit` command.



> * git config --global user.email "[you@example.com](mailto:you@example.com)"
> * git config --global user.name "Your Name"

Yes, well, as the name implies that's a _global_ config. So if you haven't done it, then yes it's needed. For most `git` users those configs are stored in their `$HOME/.gitconfig` already.

**Comment by ferdnyc** [comment]
The example is admittedly extremely contrived, as minimal, verifiable examples often are, but the situation where I originally encountered this was anything but. The same issue will come up if, on a Fedora 40 system (with the `fedpkg` tool installed and developer credentials set up for `src.fedoraproject.org`), you were to run the following:

```bash
fedpkg clone doxygen
cd doxygen
spectool -g doxygen.spec
fedpkg --release f40 build
```
...Which is what I actually did do, and got a packaged doxygen with the git SHA set to the HEAD of the `https://src.fedoraproject.org/rpms/doxygen/` repo.

**Comment by albert-github** [comment]
I've just pushed a proposed patch, pull request #10996

**Comment by albert-github** [comment]
Code has been integrated in master on GitHub (please don't close the issue as this will be done at the moment of an official release).

**Comment by ferdnyc** [comment]
@albert-github As an alternative solution to this, I belatedly discovered that the `git` command takes a top-level `--git-dir` flag, which according to the general `git(1)` manpage, will...:

>            Set the path to the repository (".git" directory). This can also be
>            controlled by setting the GIT_DIR environment variable. It can be
>            an absolute path or relative path to current working directory.
> 
>            Specifying the location of the ".git" directory using this option
>            (or GIT_DIR environment variable) turns off the repository
>            discovery that tries to find a directory with ".git" subdirectory
>            (which is how the repository and the top-level of the working tree
>            are discovered), and tells Git that you are at the top level of the
>            working tree. If you are not at the top-level directory of the
>            working tree, you should tell Git where the top-level of the
>            working tree is, with the --work-tree=<path> option (or
>            GIT_WORK_TREE environment variable)

So, another approach would be to simply make all of the `git` calls with `--git-dir=${_working_dir}/.git` (or just `--git-dir=.git`) in the argument list, and it won't go wandering if that path isn't found.

(I tested, and it does work even in a worktree when the `--git-dir` is actually a file named `.git`.)

If there's nothing there, the git command will output

> fatal: not a git repository '.git'

and abort with a non-zero exit code.

**Comment by albert-github** [comment]
@ferdnyc 
Interesting and would probably also have worked (or is even more "git"-like). Problem is that git is a bit of a black box with a lot of possible options (which are not always clear). I think that the current implementation is working fine, so we should leave it.

**Comment by ferdnyc** [comment]
@albert-github 

Makes sense, I just wanted to throw it out there. I agree the current fix works fine as-is, too.

**Comment by doxygen** [comment]
This issue was previously marked 'fixed but not released',
which means it should be fixed in doxygen version 1.12.0.
Please verify if this is indeed the case. Reopen the
issue if you think it is not fixed and please include any additional information
that you think can be relevant (preferably in the form of a self-contained example).
